src/Node.py

```diff
@@ -249,6 +304,22 @@ def get_messages(self):
             messages = self.storage.messages.copy()
         return messages
 
+    def update_balloting_state(self, ballot):
+        """
+        Update the node's internal state based on quorum threshold checks for any of the prescribed state
+        """
+        if self.check_quorum_threshold_for_field(ballot, 'voted'):
+            self.balloting_state['voted'][ballot.value.hash] = ballot
+            log.node.info('Quorum threshold met for \'voted\' for ballot %s. Updated state', ballot)
+
+        if self.check_quorum_threshold_for_field(ballot, 'accepted'):
+            self.balloting_state['accepted'][ballot.value.hash] = ballot
```

Comment:
Add a check that checks if the ballot.value.hash is already there. This could happen if a ballot is accepted and then a new ballot with the same value but a higher counter gets voted. In this case we keep the one with the higher counter